src/v/cloud_storage/segment_meta_cstore.cc

```diff
@@ -10,49 +10,554 @@
 
 #include "cloud_storage/segment_meta_cstore.h"
 
+#include "cloud_storage/types.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+#include "model/timestamp.h"
+#include "utils/delta_for.h"
+
+#include <bitset>
+#include <functional>
+#include <tuple>
+
 namespace cloud_storage {
 
+using int64_delta_alg = details::delta_delta<int64_t>;
+using int64_xor_alg = details::delta_xor;
+// Column for monotonically increaseing data
+using counter_col_t = segment_meta_column<int64_t, int64_delta_alg>;
+// Column for varying data
+using gauge_col_t = segment_meta_column<int64_t, int64_xor_alg>;
+
+/// Samping rate of the indexer inside the column store, if
+/// sampling_rate == 1 every row is indexed, 2 - every secod row, etc
+/// The value 8 with max_frame_size set to 64 will give us 8 hints per
+/// frame (frame has 64 rows). There is no measureable difference between
+/// value 8 and smaller values.
+static constexpr uint32_t sampling_rate = 8;
+
+// Sampling rate shold be proportional to max_frame_size so we will
+// sample first row of every frame.
+static_assert(
+  gauge_col_t::max_frame_size % sampling_rate == 0, "Invalid sampling rate");
+
+enum class segment_meta_ix {
+    is_compacted,
+    size_bytes,
+    base_offset,
+    committed_offset,
+    base_timestamp,
+    max_timestamp,
+    delta_offset,
+    ntp_revision,
+    archiver_term,
+    segment_term,
+    delta_offset_end,
+    sname_format,
+};
+
+namespace details {
+template<class T>
+void commit_one(T&& tx) {
+    if (tx.has_value()) {
+        tx->commit();
+    }
+}
+
+template<class... Args>
+void commit_all_impl(Args&&... args) {
+    (commit_one(args), ...);
+}
+
+template<class... Args>
+void commit_all(std::tuple<Args...>&& tup) {
+    std::apply(commit_all_impl<Args...>, std::move(tup));
+}
```

Comment:
if you are taking forwarding references in these, why not forward the value category too?

Reply:
with changes in the last push we need to move the value